crypto/bn/bn_mont.c

```diff
@@ -20,29 +20,43 @@
 #define MONT_WORD               /* use the faster word-based algorithm */
 
 #ifdef MONT_WORD
-static int BN_from_montgomery_word(BIGNUM *ret, BIGNUM *r, BN_MONT_CTX *mont);
+static int bn_from_montgomery_word(BIGNUM *ret, BIGNUM *r, BN_MONT_CTX *mont);
 #endif
 
 int BN_mod_mul_montgomery(BIGNUM *r, const BIGNUM *a, const BIGNUM *b,
                           BN_MONT_CTX *mont, BN_CTX *ctx)
+{
+    int ret = bn_mul_mont_fixed_top(r, a, b, mont, ctx);
+
+    bn_correct_top(r);
+    bn_check_top(r);
+
+    return ret;
+}
+
+int bn_mul_mont_fixed_top(BIGNUM *r, const BIGNUM *a, const BIGNUM *b,
+                          BN_MONT_CTX *mont, BN_CTX *ctx)
 {
     BIGNUM *tmp;
     int ret = 0;
-#if defined(OPENSSL_BN_ASM_MONT) && defined(MONT_WORD)
     int num = mont->N.top;
 
+#if defined(OPENSSL_BN_ASM_MONT) && defined(MONT_WORD)
     if (num > 1 && a->top == num && b->top == num) {
         if (bn_wexpand(r, num) == NULL)
             return 0;
         if (bn_mul_mont(r->d, a->d, b->d, mont->N.d, mont->n0, num)) {
             r->neg = a->neg ^ b->neg;
             r->top = num;
-            bn_correct_top(r);
+            r->flags |= BN_FLG_FIXED_TOP;
             return 1;
         }
```

Comment:
Formally speaking this, i.e. case when bn_mul_mont returns 0, calls for following question. Or more specifically in cases when it *chooses* to return 0, because it was measured to be slower than alternative. And question is should one prioritize security and perform calculations in bn_mul_mont even if they are known to be slower?

For reference, what is the potential leakage from the alternative? [And how measurable can it be?] It should be noted that if inputs are with fixed tops, the variation at least won't be computational. This is because BN_mul and BN_sqr are perfectly capable of processing zero-padded inputs, so that *computations* will take constant time, they will just correct the top for the result. Which is then immediately effectively readjusted in following reduction, yielding zero-padded output (also in constant computational time). On other words the only variation is when top is corrected and padding is added back. Both can be made constant-time... Readjustment can be more considered more dangerous, because it's done with memset. What makes it more dangerous is that you don't know if compiler with call memset or generate equivalent code inline. At least in latter case attacker won't be able to trigger variations by flushing code from cache, while depending on length memset can be tricked into it...

Just in case, again, this applies to platforms **other than** most popular ones. Well, it also applies to platforms without assembly Montgomery multiplication module, or no-asm cases.

Reply:
Just in case, above remark is specifically about what happens when bn_mul_mont **could** have executed, i.e. when both inputs have same bn->tops as modulus. Cases when this condition doesn't hold are of course different matter. So it's also about caller avoiding these cases by passing zero-padded inputs.